### PR TITLE
add sam support

### PR DESCRIPTION
**Pull request by Yogpod**
fixes 
https://github.com/BadgerCode/tttdamagelogs/issues/39

kinda deals with
https://github.com/BadgerCode/tttdamagelogs/issues/6

now that i think of it i probably should've used the `next version` branch instead of master
tell me if you spot any mistakes i tested this very briefly

### CODE REVIEW

**Comment by BadgerCode** [comment]
Hey! Thanks for this.
I hadn't heard of the "SAM" admin system before.

I'll take a look over it today/tomorrow.
I won't be able to test the SAM stuff as I don't own the mod.

File: lua/damagelogs/shared/events.lua [hunk 1]
```diff
@@ -37,8 +37,9 @@ function Damagelog:AddEvent(event, f)
             end
         end
 
+        infos = util.Compress(util.TableToJSON(infos))
         net.Start("DL_RefreshDamagelog")
-        net.WriteTable(infos)
+        net.WriteData(infos, #infos)
```
Comment (BadgerCode) [inline lua/damagelogs/shared/events.lua:42]: This should write the length of the data first, so that the client knows how much data to read.

The value of the `len` parameter in `net.Receive` is longer than the actual payload
image


Server side
```lua
net.WriteUInt(string.len(infos), 32)
net.WriteData(infos, string.len(infos))
```


Then on the client's end
```lua
local dataSize = net.ReadUInt(32)
local tbl = util.JSONToTable(util.Decompress(net.ReadData(dataSize)))
```

File: lua/damagelogs/config/config.lua [hunk 2]
```diff
@@ -49,36 +50,36 @@ REQUIRES ULX ! If you are using ServerGuard, set this to 0 (it will use ServerGu
...
+Damagelog.Autoslay_DefaultReason3 = "2x RDM"
+Damagelog.Autoslay_DefaultReason4 = "Attempted Mass"
+Damagelog.Autoslay_DefaultReason5 = "Mass RDM"
+Damagelog.Autoslay_DefaultReason6 = "Super Mass"
+Damagelog.Autoslay_DefaultReason7 = "Ghosting"
+Damagelog.Autoslay_DefaultReason8 = "Hacking"
+Damagelog.Autoslay_DefaultReason9 = "Prop kill"
+Damagelog.Autoslay_DefaultReason10 = "Consistent RDM"
+Damagelog.Autoslay_DefaultReason11 = "Yolo"
+Damagelog.Autoslay_DefaultReason12 = "Yolo"
```
Comment (BadgerCode) [inline lua/damagelogs/config/config.lua:68]: I think most of these are better than the original default reasons.
Could we replace the two "Yolo" reasons with something else?

`Damagelog.Autoslay_DefaultReason` will need re-adding, as this is used in some parts of the code

File: lua/damagelogs/config/config.lua [hunk 3]
```diff
@@ -49,36 +50,36 @@ REQUIRES ULX ! If you are using ServerGuard, set this to 0 (it will use ServerGu
...
+Damagelog.Ban_DefaultReason3 = "2x RDM and leave"
+Damagelog.Ban_DefaultReason4 = "Attempted Mass and leave"
+Damagelog.Ban_DefaultReason5 = "Mass RDM"
+Damagelog.Ban_DefaultReason6 = "Super Mass"
+Damagelog.Ban_DefaultReason7 = "Ghosting"
+Damagelog.Ban_DefaultReason8 = "Hacking"
+Damagelog.Ban_DefaultReason9 = "Consistent RDM"
+Damagelog.Ban_DefaultReason10 = "Yolo"
+Damagelog.Ban_DefaultReason11 = "Yolo"
+Damagelog.Ban_DefaultReason12 = "Yolo"
```
Comment (BadgerCode) [inline lua/damagelogs/config/config.lua:82]: Could the Yolos be replaced with something else?

File: lua/damagelogs/config/config.lua [hunk 4]
```diff
@@ -88,7 +89,7 @@ Damagelog.UseWorkshop = true
 -- Force a language - When empty use user-defined language
 Damagelog.ForcedLanguage = ""
 -- Allow reports even with no staff online
-Damagelog.NoStaffReports = false
+Damagelog.NoStaffReports = true
```
Comment (BadgerCode) [inline lua/damagelogs/config/config.lua:92]: Was this change intentional?
I try to leave the default config the way it is to avoid causing confusing when people upgrade.

File: lua/damagelogs/client/tabs/rdm_manager.lua [hunk 5]
```diff
@@ -320,13 +326,17 @@ local function TakeAction()
         slaynr2:AddOption(TTTLogTranslate(GetDMGLogLang, "ReportedPlayer") .. " (" .. report.attacker_nick .. ")", function()
             if IsValid(attacker) then
                 if ulx then
-                    RunConsoleCommand("ulx", mode == 1 and "aslayid" or "ajailid", report.attacker, "0")
-                else
-                    serverguard.command.Run("raslay", false, attacker:Nick())
+                    RunConsoleCommand("ulx", mode == 1 and "aslay" or "ajail", attacker:Nick(), "0")
```
Comment (BadgerCode) [inline lua/damagelogs/client/tabs/rdm_manager.lua:329]: Can this go back to `aslayid`/`ajailid`?

I remember this command having problems in the past with players that had weird characters in their names.

File: lua/damagelogs/client/tabs/rdm_manager.lua [hunk 6]
```diff
@@ -336,13 +346,17 @@ local function TakeAction()
         slaynr2:AddOption(TTTLogTranslate(GetDMGLogLang, "TheVictim") .. " (" .. report.victim_nick .. ")", function()
             if IsValid(victim) then
                 if ulx then
-                    RunConsoleCommand("ulx", mode == 1 and "aslayid" or "ajailid", report.attacker, "0")
-                else
+                    RunConsoleCommand("ulx", mode == 1 and "aslay" or "ajail", victim:Nick(), "0")
```
Comment (BadgerCode) [inline lua/damagelogs/client/tabs/rdm_manager.lua:349]: Can this go back to `aslayid`/`ajailid`?

I remember this command having problems in the past with players that had weird characters in their names.

File: lua/damagelogs/client/tabs/rdm_manager.lua [hunk 7]
```diff
@@ -1058,8 +1072,10 @@ function PANEL:SetPlayer(reported, ply, steamid, report)
     self.Button.DoClick = function(panel)
         if IsValid(ply) then
             if ulx then
-                RunConsoleCommand("ulx", mode == 1 and "aslayid" or "ajailid", ply:SteamID(), tostring(self.NumSlays), self.CurrentReason)
-            else
+                RunConsoleCommand("ulx", mode == 1 and "aslay" or "ajail", ply:Nick(), tostring(self.NumSlays), self.CurrentReason)
```
Comment (BadgerCode) [inline lua/damagelogs/client/tabs/rdm_manager.lua:1075]: Can this go back to `aslayid`/`ajailid`?

I remember this command having problems in the past with players that had weird characters in their names.

File: lua/damagelogs/server/discord.lua [hunk 8]
```diff
@@ -11,6 +11,10 @@ local emitOnlyWhenAdminsOffline = Damagelog.DiscordWebhookMode == POST_MODES.WHE
 local limit = 5
 local reset = 0
 
+local use_chttp = pcall(require, "chttp")
+if use_chttp then
+    HTTP = CHTTP
+end
```
Comment (BadgerCode) [inline lua/damagelogs/server/discord.lua:17]: This should fix #6 
I will need to update the wiki with install instructions.
I shall test this

**Review by BadgerCode — CHANGES_REQUESTED** [review]
Looks good!
There are a few minor things I've commented on.

I shall need to do some testing.
I won't be able to test the SAM stuff as I do not have the addon.

File: lua/damagelogs/client/tabs/rdm_manager.lua (hunk 5 above)
Comment (Yogpod) [reply lua/damagelogs/client/tabs/rdm_manager.lua]: done

File: lua/damagelogs/client/tabs/rdm_manager.lua (hunk 6 above)
Comment (Yogpod) [reply lua/damagelogs/client/tabs/rdm_manager.lua]: done

File: lua/damagelogs/client/tabs/rdm_manager.lua (hunk 7 above)
Comment (Yogpod) [reply lua/damagelogs/client/tabs/rdm_manager.lua:1075]: done

File: lua/damagelogs/config/config.lua (hunk 2 above)
Comment (Yogpod) [reply lua/damagelogs/config/config.lua]: done

File: lua/damagelogs/config/config.lua (hunk 4 above)
Comment (Yogpod) [reply lua/damagelogs/config/config.lua]: it wasn't

File: lua/damagelogs/server/discord.lua (hunk 8 above)
Comment (Yogpod) [reply lua/damagelogs/server/discord.lua:17]: awesome

File: lua/damagelogs/shared/events.lua (hunk 1 above)
Comment (Yogpod) [reply lua/damagelogs/shared/events.lua:43]: done

**Review by Yogpod — COMMENTED** [review]
done

File: lua/damagelogs/config/config.lua (hunk 3 above)
Comment (Yogpod) [reply lua/damagelogs/config/config.lua]: done

**Comment by Yogpod** [comment]
wow sorry for probably spamming your email, never done a review thing so im not sure exactly how it works

File: lua/damagelogs/shared/events/equipment.lua [hunk 9]
```diff
@@ -28,7 +28,7 @@ function event:TTTEquipmentUse(ply, equipment, info)
         [EVENT_DETAILS.UserName] = ply:Nick(),
         [EVENT_DETAILS.UserRole] = ply:GetRole(),
         [EVENT_DETAILS.UserSteamID] = ply:SteamID(),
-        [EVENT_DETAILS.EquipmentClass] = equipment:GetClass(),
+        [EVENT_DETAILS.EquipmentClass] = type(equipment) == "Weapon" and equipment:GetClass() or equipment,
```
Comment (BadgerCode) [inline lua/damagelogs/shared/events/equipment.lua:31]: Is this to allow strings to be passed to this instead of entities?
It might be safer to do
`IsEntity(equipment) and equipment:GetClass() or equipment`

As it's possible that non-weapon entities could use this hook (e.g. planted bombs, custom NPCs, grenade entities)

**Comment by BadgerCode** [comment]
No worries!

## Discord integration
* ✅ Discord notifications work when using a proxy & not using CHTTP
* ✅ Discord notifications work when using a proxy & using CHTTP
* ✅ Discord notifications work when not using a proxy & using CHTTP

## Aslay
* ✅ Correct output from cslays
* ✅ Correct output from cslaysid
* ✅ Can aslay a user
* ✅ Can change the slays for a user who already has some
* ✅ Correct output when a player is slain

## General
* ✅ View logs of current round
* ✅ View logs of previous round
* ✅ View logs of a round from a different map
* ✅ Report a player & have them respond
* ✅ Be reported & respond

**Review by BadgerCode — APPROVED** [review]
Looks great!

Thanks a lot for fixing the Discord & cslays issue, and adding support for SAM!
I'll try get this release this week.

**Comment by Yogpod** [comment]
yea np
